trieste/models/gpflux/sampler.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2021 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import tensorflow as tf
+from gpflux.layers import GPLayer, LatentVariableLayer
+from gpflux.models import DeepGP
+from gpflux.sampling.sample import Sample
+
+from ...types import TensorType
+from ...utils import DEFAULTS
+from ..interfaces import (
+    ProbabilisticModel,
+    ReparametrizationSampler,
+    TrajectoryFunction,
+    TrajectorySampler,
+)
+from .interface import GPfluxPredictor
+
+
+def sample_consistent_lv_layer(layer: LatentVariableLayer) -> Sample:
+    class SampleLV(Sample):  # type: ignore[misc]
+        def __call__(self, X: TensorType) -> tf.Tensor:
+            sample = layer.prior.sample()
+            batch_shape = tf.shape(X)[:-1]
+            sample_rank = tf.rank(sample)
+            for _ in range(len(batch_shape)):
+                sample = tf.expand_dims(sample, 0)
+            sample = tf.tile(
+                sample, tf.concat([batch_shape, tf.ones(sample_rank, dtype="int32")], -1)
+            )
+            return layer.compositor([X, sample])
+
+    return SampleLV()
+
+
+def sample_dgp(model: DeepGP) -> TrajectoryFunction:
+    function_draws = []
+    for layer in model.f_layers:
+        if isinstance(layer, GPLayer):
+            function_draws.append(layer.sample())
+        elif isinstance(layer, LatentVariableLayer):
+            function_draws.append(sample_consistent_lv_layer(layer))
+        else:
+            raise NotImplementedError(f"Sampling not implemented for {type(layer)}")
+
+    class ChainedSample(Sample):  # type: ignore[misc]
+        def __call__(self, X: TensorType) -> tf.Tensor:
+            for f in function_draws:
+                X = f(X)
+            return X
+
+    return ChainedSample().__call__
+
+
+class DeepGaussianProcessTrajectorySampler(TrajectorySampler[GPfluxPredictor]):
+    r"""
```

Comment:
docstring is in conflict with type stated in the class, is this valid only for DeepGaussianProcess, or for any subclass of GPfluxPredictor?

Reply:
(though if you import DeepGaussianProcess you'll get circular imports...)

Reply:
It should be valid for any GPfluxPredictor with a gpflux.models.DeepGP model. This is essentially the definition of DeepGaussianProcess though, although as you say we need to avoid circular imports. If you have any better ideas I'm happy to try them out :)

Reply:
I've tried to clarify this a bit in the docstring now as well

Reply:
huh, this part is a bit messy, if we look to GPflow there e.g. `IndependentReparametrizationSampler` states `ProbabilisticModel` even though it seems its meant to be used with GPflow models - perhaps that one should be moved up to a parent folder if any prob model can use it...

otherwise, GPflowPredictor seems fine for now, although its a bit weird that it actually only allows DeepGaussianProcess, so why not specifying that immediately (in which case the only solution is to move this to models.py)... other types of models you have implemented previously, they were also deepGP class? we could have different interfaces here for the same gpflux class?